### PR TITLE
global: local file identifiers

### LINKED ISSUE #5: dereferencing local files
Problem:

``` console
$ workon domapping
$ cd dojson/contrib/marc21/schemas
$ domapping schema_to_mapping ./marc21.json
[...]
jsonschema.exceptions.RefResolutionError: unknown url type: bd00x
```

Input:

``` console
$ cat marc21.json 
{
    "$schema": "http://json-schema.org/draft-04/schema#",
    "anyOf": [
        {"$ref": "bd00x#"},
        {"$ref": "bd01x09x#"},
        {"$ref": "bd1xx#"},
        {"$ref": "bd20x24x#"},
        {"$ref": "bd25x28x#"},
        {"$ref": "bd3xx#"},
        {"$ref": "bd4xx#"},
        {"$ref": "bd5xx#"},
        {"$ref": "bd6xx#"},
        {"$ref": "bd70x75x#"},
        {"$ref": "bd76x78x#"},
        {"$ref": "bd80x83x#"},
        {"$ref": "bd84188x#"}
    ]
}
```

Possible solutions:
- use reasonable URIs?
- join files together?  (is `anyOf` nicely supported?)
- fix file discovery in case of local files?

### PR DESCRIPTION
**Pull request by nharraud**
closes #5

### CODE REVIEW

**Comment by nharraud** [comment]
@JavierDelgadoFernandez @tiborsimko here is the PR. Can you tell be if this can generate your mappings?
I also added type guessing from enum when they contain only "string" values.

File: domapping/mapping.py [hunk 1]
```diff
@@ -312,6 +316,24 @@ def __gen_type_properties(json_schema, path, resolver, config, es_mapping):
     return es_mapping
 
 
+def __guess_enum_type(enum_array, path):
```
Comment (jirikuncar) [inline domapping/mapping.py]: @nharraud do not use double underscore for making function "private".

## 

https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles

Comment (JavierDelgadoFernandez) [inline domapping/mapping.py]: Thanks a lot @nharraud. I was testing the tool right now and the result for this file [CMS Questionnaire](https://github.com/cernanalysispreservation/analysis-preservation.cern.ch/blob/master/cap/modules/experiments/static/jsonschemas/cms/cms_questionnaire_schema.json). I needed to remove ``aditionalProperties`` from the schema (Only two lines out of 1200). The result looks like:
```json
{
    "date_detection": true,
    "_all": {
        "enable": true
    },
    "properties": {
        "general": {
            "type": "object",
            "properties": {
                "faq_awareness": {
                    "type": "string"
                },
                "meetings_awareness": {
                    "type": "string"
                },
                "forum_awareness": {
                    "type": "string"
                },
                "forum_subscription": {
                    "type": "string"
                },
                "past_interaction": {
                    "type": "string"
                }
            }
        },
        "multivariate": {
            "type": "object",
            "properties": {
                "correlation_studies": {
                    "type": "object",
                    "properties": {
                        "name": {
                            "type": "string"
                        },
                        "comments": {
                            "type": "string"
                        }
                    }
                },
(...)
                "usage": {
                    "type": "string"
                },
                "bcgrnd_shape": {
                    "type": "object",
                    "properties": {
                        "name": {
                            "type": "string"
                        },
                        "comments": {
                            "type": "string"
                        }
                    }
                }
            }
        },
        "feedback": {
            "type": "object",
            "properties": {
                "subject_to_change": {
                    "type": "object",
                    "properties": {
                        "name": {
                            "type": "string"
                        },
                        "comments": {
                            "type": "string"
                        }
                    }
                },
                "feedback": {
                    "type": "string"
                },
                "additional_comments": {
                    "type": "string"
                }
            }
        }
    },
    "numeric_detection": true
}
```

Comment (jirikuncar) [inline domapping/mapping.py]: @JavierDelgadoFernandez you might want to move this inline comment as global comment ;)

**Comment by JavierDelgadoFernandez** [comment]
Thanks a lot @nharraud. I was testing the tool right now and here is the result for this file [CMS Questionnaire](https://github.com/cernanalysispreservation/analysis-preservation.cern.ch/blob/master/cap/modules/experiments/static/jsonschemas/cms/cms_questionnaire_schema.json). I needed to remove `aditionalProperties` from the schema (Only two lines out of 1200). The result looks like:

``` json
{
    "date_detection": true,
    "_all": {
        "enable": true
    },
    "properties": {
        "general": {
            "type": "object",
            "properties": {
                "faq_awareness": {
                    "type": "string"
                },
                "meetings_awareness": {
                    "type": "string"
                },
                "forum_awareness": {
                    "type": "string"
                },
                "forum_subscription": {
                    "type": "string"
                },
                "past_interaction": {
                    "type": "string"
                }
            }
        },
        "multivariate": {
            "type": "object",
            "properties": {
                "correlation_studies": {
                    "type": "object",
                    "properties": {
                        "name": {
                            "type": "string"
                        },
                        "comments": {
                            "type": "string"
                        }
                    }
                },
(...)
                "usage": {
                    "type": "string"
                },
                "bcgrnd_shape": {
                    "type": "object",
                    "properties": {
                        "name": {
                            "type": "string"
                        },
                        "comments": {
                            "type": "string"
                        }
                    }
                }
            }
        },
        "feedback": {
            "type": "object",
            "properties": {
                "subject_to_change": {
                    "type": "object",
                    "properties": {
                        "name": {
                            "type": "string"
                        },
                        "comments": {
                            "type": "string"
                        }
                    }
                },
                "feedback": {
                    "type": "string"
                },
                "additional_comments": {
                    "type": "string"
                }
            }
        }
    },
    "numeric_detection": true
}
```

**Comment by nharraud** [comment]
Ready for review.

File: examples/schema.json [hunk 2]
```diff
@@ -8,7 +7,10 @@
             "format": "custom-date"
         },
         "company": {"type": "string"},
-        "private_field": {"type": "string"}
+        "private_field": {"type": "string"},
+        "external_test": {
+            "$ref": "subschema.json#/definitions/externaltype"
```
Comment (switowski) [inline examples/schema.json]: This file shouldn't be commited.

File: domapping/mapping.py [hunk 3]
```diff
@@ -217,9 +217,13 @@ def __gen_type_properties(json_schema, path, resolver, config, es_mapping):
     if not json_type:
         if 'properties' in json_schema:
             json_type = 'object'
+        elif 'enum' in json_schema:
+            json_type = _guess_enum_type(json_schema['enum'], path)
         else:
-            # FIXME: handle enums with no type
-            raise JsonSchemaSupportError('Schema has no "type" field', path)
+            raise UnknownFieldTypeError(
+                'Schema field type cannot be guessed. Onlly fields with "type"'
```
Comment (switowski) [inline domapping/mapping.py]: Typo

**Comment by nharraud** [comment]
@lnielsen I will separate this in 2 PR as requested by @switowski

**Comment by nharraud** [comment]
Split done. Enum PR is #8

File: domapping/cli.py [hunk 4]
```diff
@@ -41,6 +44,7 @@ def cli():
 
 @cli.command('schema_to_mapping')
 @click.argument('schema', type=click.File('r'), default='-')
+# @click.argument('load-schemas', type=click.Path(exists=True))
```
Comment (lnielsen) [inline domapping/cli.py]: Remove ?

Comment (nharraud) [inline domapping/cli.py]: Done

File: setup.py [hunk 5]
```diff
@@ -42,6 +42,7 @@
     'pytest-cov>=1.8.0',
     'pytest-pep8>=1.0.6',
     'pytest>=2.8.0',
+    'responses>=0.5.1'
```
Comment (lnielsen) [inline setup.py:45]: Normally we've been using HTTPretty for request mocking - any particular reason for using `responses` library?

**Comment by lnielsen** [comment]
@switowski If you drop a :ship: in the comments, I'll merge.

File: setup.py (hunk 5 above)
Comment (nharraud) [inline setup.py:45]: Yes I learned that after choosing `response`. I wanted to use HTTPretty too at first but then I saw this: `HTTPretty is not supporting python3 officially. You will notice that the travis build for python 3 might be broken, and while pull requests fixing py3 support are most welcome, it is still not official at least for now.` https://github.com/gabrielfalcao/HTTPretty

**Comment by switowski** [comment]
looks good, :ship:

**Pull request merged**